File: arrays-and-strings/1.1.js
// Solution with additional data structures

function allUnique(str) {
  var charsSeen = {};

  for (var i = 0; i < str.length; i++) {
    var char = str.charAt(i);
    if (charsSeen[char]) === true {
      return false;
    } else {
      charsSeen[char] = true;
    }
  }

  return true;
}

// Solution without additional data structures

function allUniqueSmall(str) {
  for (var i = 0; i < str.length; i++) {
    var base = str.charAt(i);

    for (var j = i + 1; j < str.length; j++) {
      var check = str.charAt(j);
      if (base === check) {
        return false;
      }
    }
  }

  return true;
}

File: arrays-and-strings/1.1.py
# Solutions with additional data structures

def all_unique(string):
    chars_seen = set()
    for char in string:
        if char in chars_seen:
            return False
        else:
            chars_seen.add(char)

    return True

# Solutions without additional data structures

def all_unique_small(string):
    for i in range(len(string)):
        for j in range(i + 1, len(string)):
            if string[i] == string[j]:
                return False

    return True

def all_unique_small_sorting(string):
    sorted_string = sorted(string)
    for i in range(len(string) - 1):
        if sorted_string[i] == sorted_string[i + 1]:
            return False

    return True
